### DISCUSSION TITLE
<Tab> key no longer opens directories, only previews, after recent update

### DISCUSSION THREAD

**Discussion by haakonfp** [nvim-tree/nvim-tree.lua#2027] [Q&A] [answered]
Hello!

I've normally used ``<Tab>`` to preview files and open/close directories in the tree.
However, after I recently ran a ``:PackerUpdate``, this no longer works.

Instead, ``<Tab>`` only allows me to preview files, and does nothing with directories.
Here is my current configuration:

```
require('nvim-tree').setup({
  sort_by = 'case_sensitive',
  view = {
    adaptive_size = true,
    mappings = {
      list = {
        { key = 'u', action = 'dir_up' }
      },
    },
  },
  renderer = {
      group_empty = true,
  },
  filters = {
      dotfiles = true,
  },
  open_on_tab = true,
})
```

When I tried remapping ``<Tab>`` to some random action, like ``tabnew``, it actually opens directories again (but also opens new tabs instead of previewing files, which is not what I want).

How would I remap this to get my original functionality of previewing files and opening/closing directories?

**Comment by alex-courtis** [comment, marked as answer]
> When I tried remapping `<Tab>` to some random action, like `tabnew`, it actually opens directories again (but also opens new tabs instead of previewing files, which is not what I want).
> 
> How would I remap this to get my original functionality of previewing files and opening/closing directories?

That was actually a bug; "preview" would open file when it was on a directory, in this case expanding the directory.

Here's a function you can map to achieve this behaviour via action_cb:

```lua
local function preview_or_toggle_dir(node)
  local api = require("nvim-tree.api")

  if node.type == "file" or node.type == "link" then
    api.node.open.preview(node)
  else
    api.node.open.edit(node)
  end
end
```

P.S. please see https://github.com/nvim-tree/nvim-tree.lua/wiki/Migrating-To-on_attach to migrate your mappings to a simple `on_attach` function, the new mechanism.

**Reply by haakonfp** [reply]
Perfect, thank you! Will try to migrate to ``on_attach`` as well, thx
